pkg/etcd/migrate.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	clientV3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+	"go.uber.org/zap"
+)
+
+const (
+	// cdcMetaVersion is hard code value indicate the metaVersion of TiCDC
+	cdcMetaVersion          = 1
+	etcdSessionTTL          = 10
+	noMetaVersion           = -1
+	migrateLogsWarnDuration = 5 * time.Second
+	migrationCampaignKey    = "ticdc-migration"
+)
+
+type keys map[string]string
+
+func (k keys) addPair(old, new string) {
+	k[old] = new
+}
+
+type migrator struct {
+	clusterID      string
+	oldMetaVersion int
+	newMetaVersion int
+	metaVersionKey string
+	// cdc old owner key
+	oldOwnerKey string
+	// etcd client
+	cli *Client
+	// all keyPrefixes needed to be migrated or update
+	// map from oldKeyPrefix to newKeyPrefix
+	keyPrefixes       keys
+	etcdNoMetaVersion bool
+}
+
+// Note: we do not use etcd transaction to migrate key
+// as it has the maximum operation limit in a single transaction.
+// So we use a double check mechanism to make sure the migration is complete.
+// 1. check and put metaVersion
+// 2. campaign old owner
+// 3. update keys
+// 4. update metaVersion
+func (m migrator) migrate(ctx context.Context) error {
+	// 1.1 check metaVersion, if the metaVersion in etcd does not match
+	// m.oldMetaVersion, it means that someone has migrated the meta data
+	metaVersion, err := getMetaVersion(ctx, m.cli, m.clusterID)
+	if err != nil {
+		log.Error("get meta version failed, etcd meta data migration failed", zap.Error(err))
+		return errors.Trace(err)
+	}
+
+	if metaVersion > m.newMetaVersion {
+		log.Panic("meta version in etcd is greater than the meta version in TiCDC",
```

Comment:
Will this be triggered under normal circumstances?

Reply:
It will never happen in a normal circumstance.